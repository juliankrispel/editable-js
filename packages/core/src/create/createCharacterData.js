// @flow

import type { RawBlock, CharacterData } from '../types'

export default function createCharacterData(block: RawBlock): Array<CharacterData> {
  if (block.text != null) {
    return block.text.split('').map((_, index) => {
      const data = {
        marks: []
      }

      if (block.characterRanges != null) {
        block.characterRanges.filter(range => {
          return range.offset <= index &&
          range.offset + range.length > index
        }).forEach(range => {
          if (Array.isArray(range.marks)) {
            data.marks = data.marks.concat(range.marks)
          }

          if (range.entity != null) {
            data.entity = range.entity
          }
        })
      }

      return data
    })
  }
  return []
}
