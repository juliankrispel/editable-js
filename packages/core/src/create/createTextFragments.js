// @flow
import type { TextFragment, Block, EntityMap } from '../types'
import hasEqualCharacterData from '../hasEqualCharacterData'

export default function createTextFragments(block: Block, entityMap: EntityMap): Array<TextFragment> {
  return block.characterData.reduce(
    (acc, data, index) => {
      if (acc.length < 1) {
        return [{
          marks: data.marks,
          entity: data.entity,
          offset: index,
          text: block.text[index]
        }]
      } else {
        const lastFragment = acc[acc.length - 1]
        if (hasEqualCharacterData(lastFragment, data)) {
          return acc.slice(0, -1).concat([{
            ...lastFragment,
            text: lastFragment.text + block.text[index]
          }])
        } else {
          return [
            ...acc,
            {
              marks: data.marks,
              entity: data.entity,
              offset: index,
              text: block.text[index]
            }
          ]
        }
      }
    },
    []
  ).map(fragment => ({
    ...fragment,
    entity: fragment.entity != null ? entityMap[fragment.entity] : null
  }))
}
