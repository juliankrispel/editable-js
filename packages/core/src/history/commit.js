// @flow

import produce, { applyPatches } from 'immer'

import undo from './undo'
import type { EditorState } from '../types'

const mergeLastCommit = (editorState: EditorState) => {
  const prevEditorState = undo(undo(editorState, true), true)
  let changes = { forward: [], reverse: [] }
  const lastTwoChanges = editorState.changes.slice(-2)

  const newEditorState = produce(
    prevEditorState,
    draft => {
      lastTwoChanges.forEach(change => {
        // console.log('change', JSON.stringify(change))
        applyPatches(draft, change.forward)
      })
    }, (forward, reverse) => {
      changes = {forward, reverse}
    }
  )

  return produce(
    newEditorState,
    draft => {
      draft.changes.pop()
      draft.changes[draft.changes.length - 1] = changes
      draft.changeIndex = draft.changes.length
    }, () => {}
  )
}

const commit = (
  editorState: EditorState,
  update: (EditorState, ...args:Array<any>) => void,
  ...rest: Array<any>
) => {
  let changes = { forward: [], reverse: [] }

  const { lastCommitted } = editorState

  let newEditorState = produce(
    editorState,
    draft => {
      update(draft, ...rest)
    },
    (forward, reverse) => {
      changes = { forward, reverse }
    }
  )

  newEditorState = produce(
    newEditorState,
    draft => {
      const { changeIndex } = draft

      draft.changeIndex++

      if (draft.changes.length > changeIndex) {
        draft.changes = draft.changes.slice(0, changeIndex - 1)
      }

      draft.changes.push(changes)

      draft.lastCommitted = update.name
    }
  )

  if (lastCommitted === update.name) {
    newEditorState = mergeLastCommit(newEditorState)
  }

  return newEditorState
}

export default commit
