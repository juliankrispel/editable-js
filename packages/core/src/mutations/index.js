// @flow

import collapseBlock from './collapseBlock'
import deleteBlock from './deleteBlock'
import insertBlockAfter from './insertBlockAfter'
import insertBlockBefore from './insertBlockBefore'
import insertText from './insertText'
import replaceText from './replaceText'
import mergeBlock from './mergeBlock'
import removeRange from './removeRange'
import splitBlock from './splitBlock'
import updateBlock from './updateBlock'
import setCurrentCharacterData from './setCurrentCharacterData'
import toggleCurrentMark from './toggleCurrentMark'
import toggleBlockType from './toggleBlockType'
import collapseSelectionToEnd from './collapseSelectionToEnd'
import collapseSelectionToStart from './collapseSelectionToStart'
import updateSelection from './updateSelection'

export {
  collapseBlock,
  deleteBlock,
  insertBlockAfter,
  insertBlockBefore,
  insertText,
  replaceText,
  mergeBlock,
  removeRange,
  splitBlock,
  toggleCurrentMark,
  toggleBlockType,
  updateBlock,
  collapseSelectionToEnd,
  collapseSelectionToStart,
  updateSelection,
  setCurrentCharacterData
}
