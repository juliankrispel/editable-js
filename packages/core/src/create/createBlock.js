// @flow

import type { RawBlock, Block } from '../types'
import createId from './createId'
import createCharacterData from './createCharacterData'

export default function createBlock (block: RawBlock): Block {
  const { children, text = '', key, characterRanges, ...rest } = block

  if (process.env.NODE_ENV !== 'production' && key != null && !isNaN(Number(key))) {
    throw new Error('block key can not be a number')
  }

  const newBlock = {
    ...rest,
    text,
    characterData: createCharacterData(block),
    children: [],
    key: key || createId()
  }

  if (Array.isArray(children)) {
    newBlock.children = children.map(createBlock)
  }

  return newBlock
}
