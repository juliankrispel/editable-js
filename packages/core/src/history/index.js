import commit from './commit'
import redo from './redo'
import undo from './undo'

export {
  commit,
  redo,
  undo
}
