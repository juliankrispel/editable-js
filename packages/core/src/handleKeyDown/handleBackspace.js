// @flow

import { getBlockBefore } from '../queries'
import { isCollapsed } from '../selection'
import { removeRange } from '../mutations'

import type { EditorState } from '../types'

export default function handleBackspace (editorState: EditorState) {
  let { selection } = editorState

  if (!isCollapsed(selection)) {
    removeRange(editorState, selection)
    return
  }

  const blockBefore = getBlockBefore(editorState.content, selection.startKey)

  if (selection.startOffset > 0) {
    selection = {
      ...selection,
      startOffset: selection.startOffset - 1
    }
  } else if (blockBefore != null) {
    selection = {
      ...selection,
      startKey: blockBefore.key,
      startOffset: blockBefore.text.length
    }
  } else {
    return
  }

  removeRange(editorState, selection)
}
