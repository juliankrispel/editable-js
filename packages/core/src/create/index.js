import createBlock from './createBlock'
import createId from './createId'
import createCharacterData from './createCharacterData'
import createTextFragments from './createTextFragments'
import createEditorState from './createEditorState'

export {
  createBlock,
  createEditorState,
  createTextFragments,
  createCharacterData,
  createId
}
