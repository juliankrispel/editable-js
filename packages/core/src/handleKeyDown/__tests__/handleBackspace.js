// @flow

import { createEditorState } from '../../create'
import { commit } from '../../history'
import handleBackspace from '../handleBackspace'
import { updateSelection } from '../../mutations'

describe('handleBackspace', () => {
  const initialState = createEditorState({ blocks: [{
    text: 'Line One',
    key: '1'
  }, {
    text: 'Line Two',
    key: '2'
  }]})

  describe('when selection is not collapsed', () => {
    let newEditorState = commit(initialState, updateSelection, {
      startOffset: 4,
      startKey: '1',
      endOffset: 7,
      endKey: '2'
    })

    newEditorState = commit(newEditorState, handleBackspace)

    test('removes range', () => {
      expect(newEditorState.content).toMatchSnapshot()
    })

    test('collapses selection to start', () => {
      expect(newEditorState.selection).toMatchSnapshot()
    })
  })

  describe('when selection is collapsed', () => {
    let newEditorState = commit(initialState, updateSelection, {
      startOffset: 4,
      startKey: '1',
      endOffset: 4,
      endKey: '1'
    })

    newEditorState = commit(newEditorState, handleBackspace)

    test('removes previous character', () => {
      expect(newEditorState.content).toMatchSnapshot()
    })

    test('moves selection to prev character', () => {
      expect(newEditorState.selection).toMatchSnapshot()
    })
  })

  describe('when selection is collapsed and at start of block', () => {
    let newEditorState = commit(initialState, updateSelection, {
      startOffset: 0,
      startKey: '2',
      endOffset: 0,
      endKey: '2'
    })

    newEditorState = commit(newEditorState, handleBackspace)

    test('merges with previous block', () => {
      expect(newEditorState.content).toMatchSnapshot()
    })

    test('moves selection to end of previous block', () => {
      expect(newEditorState.selection).toMatchSnapshot()
    })
  })

  describe('when selection is collapsed and at beginning of the document', () => {
    let newEditorState = commit(initialState, updateSelection, {
      startOffset: 0,
      startKey: '1',
      endOffset: 0,
      endKey: '1'
    })

    newEditorState = commit(newEditorState, handleBackspace)

    test('editorState.content remains the same', () => {
      expect(newEditorState.content).toMatchSnapshot()
    })

    test('editorState.selection remains the same', () => {
      expect(newEditorState.selection).toMatchSnapshot()
    })
  })
})
