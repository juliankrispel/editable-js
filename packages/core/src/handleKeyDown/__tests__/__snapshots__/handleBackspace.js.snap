// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleBackspace when selection is collapsed and at beginning of the document editorState.content remains the same 1`] = `
Array [
  Object {
    "characterData": Array [
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
    ],
    "children": Array [],
    "key": "1",
    "text": "Line One",
  },
  Object {
    "characterData": Array [
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
    ],
    "children": Array [],
    "key": "2",
    "text": "Line Two",
  },
]
`;

exports[`handleBackspace when selection is collapsed and at beginning of the document editorState.selection remains the same 1`] = `
Object {
  "endKey": "1",
  "endOffset": 0,
  "startKey": "1",
  "startOffset": 0,
}
`;

exports[`handleBackspace when selection is collapsed and at start of block merges with previous block 1`] = `
Array [
  Object {
    "characterData": Array [
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
    ],
    "children": Array [],
    "key": "1",
    "text": "Line OneLine Two",
  },
]
`;

exports[`handleBackspace when selection is collapsed and at start of block moves selection to end of previous block 1`] = `
Object {
  "endKey": "1",
  "endOffset": 8,
  "startKey": "1",
  "startOffset": 8,
}
`;

exports[`handleBackspace when selection is collapsed moves selection to prev character 1`] = `
Object {
  "endKey": "1",
  "endOffset": 3,
  "startKey": "1",
  "startOffset": 3,
}
`;

exports[`handleBackspace when selection is collapsed removes previous character 1`] = `
Array [
  Object {
    "characterData": Array [
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
    ],
    "children": Array [],
    "key": "1",
    "text": "Lin One",
  },
  Object {
    "characterData": Array [
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
    ],
    "children": Array [],
    "key": "2",
    "text": "Line Two",
  },
]
`;

exports[`handleBackspace when selection is not collapsed collapses selection to start 1`] = `
Object {
  "endKey": "1",
  "endOffset": 4,
  "startKey": "1",
  "startOffset": 4,
}
`;

exports[`handleBackspace when selection is not collapsed removes range 1`] = `
Array [
  Object {
    "characterData": Array [
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
    ],
    "children": Array [],
    "key": "1",
    "text": "Lineo",
  },
]
`;
